doc/development/05_repo_manager.md

```diff
@@ -0,0 +1,202 @@
+pub fn init(&self, uris: PublicationServerUris) -> KrillResult<()> {
+    info!("Initializing repository");
+    self.access.init(uris.clone(), &self.signer)?;
+    self.content.init(&self.config.data_dir, uris)?;
+    self.content.write_repository(&self.config.repository_retention)?;
+
+    Ok(())
+}
+```
+
+Note that it writes the empty repository as part of the initialization process.
+We do this, because that gives users an opportunity to verify that the repository
+content can be accessed. In particular: expect an RRDP notification.xml and a snapshot.xml
+containing 0 entries to be published.
+
+If it is found that the repository is not set up correctly, then it can be 'cleared'
+so it can be initialized again:
+
+```rust
+/// Clear the publication server. Will fail if it still
+/// has publishers. Or if it does not exist
+pub fn repository_clear(&self) -> KrillResult<()> {
+    self.access.clear()?;
+    self.content.clear()
+}
+```
+
+
+Adding / Removing Publishers
+----------------------------
+
+To add a publisher we have to submit an `rfc8183::PublisherRequest`, and an `Actor`
+so that we attribute the change in the history of the `RepositoryAccess` component.
+The request type can be created from XML supplied by a remote publisher. The 'handle'
+is the local name that we will use for the publisher, and the XML contains a suggestion
+set by the remote party. But, it is possible (and perhaps recommended) to override
+this 'handle' in the request object with something that is unique to *us*, e.g. a UUID.
+
+```rust
+/// Adds a publisher. This will fail if a publisher already exists for the handle in the request.
+pub fn create_publisher(&self, req: rfc8183::PublisherRequest, actor: &Actor) -> KrillResult<()> {
+    let name = req.publisher_handle().clone();
+
+    self.access.add_publisher(req, actor)?;
+    self.content.add_publisher(name)
+}
+```
+
+The publisher is going to need to get RFC8183 Repository Response XML. We can get
+an `rfc8183::RepositoryResponse` which can be printed (`fmt::Display`) as XML:
+
+```rust
+/// Returns the RFC8183 Repository Response for the publisher
+pub fn repository_response(
+    &self,
+    rfc8181_uri: uri::Https,
+    publisher: &PublisherHandle,
+) -> KrillResult<rfc8183::RepositoryResponse> {
+    self.access.repository_response(rfc8181_uri, publisher)
+}
+```
+
+In order to remove a publisher we just need to tell the RepositoryManager its handle (name).
+This removes both the access information (handle, RFC8181 ID cert, ..) and all content for
+the publisher:
+
+```rust
+/// Removes a publisher and all of its content.
+pub fn remove_publisher(&self, name: PublisherHandle, actor: &Actor) -> KrillResult<()> {
+    let publisher = self.access.get_publisher(&name)?;
+    let base_uri = publisher.base_uri();
+
+    self.content.remove_publisher(&name, base_uri, &self.config.repository_retention)?;
+    self.access.remove_publisher(name, actor)
```

Comment:
Is it right to remove the content then the publisher? If the content removal succeeds but the publisher removal fails the publisher will be left behind but lacking content? Is that right?

Reply:
It's extremely unlikely. And the other way around is even worse because then we won't know anymore which content to remove.. so.. let me just make the content removal idem-potent.. i.e. we don't need to care if it failed because there was no content.